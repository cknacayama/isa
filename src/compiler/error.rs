use std::rc::Rc;

use thiserror::Error;

use crate::compiler::token::TokenKind;

use super::infer::Constr;

#[derive(Error, Debug, Clone, Copy, PartialEq, Eq)]
pub enum LexError {
    #[error("invalid character '{0}'")]
    InvalidChar(char),
}

#[derive(Error, Debug, Clone, Copy, PartialEq, Eq)]
pub enum ParseError {
    #[error("{0}")]
    LexError(#[from] LexError),
    #[error("unexpected end-of-file")]
    UnexpectedEof,
    #[error("expected '{0}'")]
    ExpectedToken(TokenKind<'static>),
    #[error("expected expression")]
    ExpectedExpr,
    #[error("expected identifier")]
    ExpectedId,
    #[error("expected type")]
    ExpectedType,
    #[error("expected pattern")]
    ExpectedPattern,
}

#[derive(Error, Debug, Clone, PartialEq, Eq)]
pub enum InferError {
    #[error("uninferable: {0}")]
    Uninferable(Constr),
    #[error("unbound identifier: {0}")]
    Unbound(Rc<str>),
}
