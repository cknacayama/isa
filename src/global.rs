use std::fmt::Display;
use std::hash::{Hash, Hasher};
use std::ops::{Deref, Range};
use std::sync::{Mutex, MutexGuard, OnceLock};

use rustc_hash::{FxHashMap, FxHashSet};

use crate::compiler::types::{TyId, TyKind};
use crate::span::SpanData;

#[derive(Clone, Copy, PartialEq, Eq, Hash)]
pub struct Symbol(u32);

#[derive(Clone, Copy, PartialEq, Eq)]
pub struct Span(u32);

#[derive(Debug, Clone, Copy)]
pub struct Ty(&'static TyKind);

#[derive(Debug, Clone, Copy)]
pub struct TySlice(&'static [Ty]);

#[derive(Debug, Clone, Copy)]
pub struct TyQuant(&'static [TyId]);

impl Deref for TySlice {
    type Target = [Ty];

    fn deref(&self) -> &'static Self::Target {
        self.0
    }
}

impl Default for TySlice {
    fn default() -> Self {
        Ty::empty_slice()
    }
}

impl Default for TyQuant {
    fn default() -> Self {
        Ty::empty_quant()
    }
}

impl Deref for TyQuant {
    type Target = [TyId];

    fn deref(&self) -> &'static Self::Target {
        self.0
    }
}

impl Hash for Ty {
    fn hash<H: Hasher>(&self, state: &mut H) {
        std::ptr::hash(self.0, state);
    }
}

impl PartialEq for Ty {
    fn eq(&self, other: &Self) -> bool {
        std::ptr::eq(self.0, other.0)
    }
}

impl Eq for Ty {
}

impl Hash for TySlice {
    fn hash<H: Hasher>(&self, state: &mut H) {
        std::ptr::hash(self.0, state);
    }
}

impl PartialEq for TySlice {
    fn eq(&self, other: &Self) -> bool {
        std::ptr::eq(self.0, other.0)
    }
}

impl Eq for TySlice {
}

impl Hash for TyQuant {
    fn hash<H: Hasher>(&self, state: &mut H) {
        std::ptr::hash(self.0, state);
    }
}

impl PartialEq for TyQuant {
    fn eq(&self, other: &Self) -> bool {
        std::ptr::eq(self.0, other.0)
    }
}

impl Eq for TyQuant {
}

impl Default for Span {
    fn default() -> Self {
        Self::zero()
    }
}

impl Span {
    #[must_use]
    pub const fn new(idx: u32) -> Self {
        Self(idx)
    }

    #[must_use]
    pub const fn index(self) -> usize {
        self.0 as usize
    }

    #[must_use]
    pub const fn zero() -> Self {
        Self(0)
    }
}

impl Default for Symbol {
    fn default() -> Self {
        Self::zero()
    }
}

impl Symbol {
    #[must_use]
    pub const fn zero() -> Self {
        Self(0)
    }

    pub fn intern(symbol: &str) -> Self {
        Env::get(|mut e| e.symbols.intern(symbol))
    }

    pub fn intern_owned(symbol: String) -> Self {
        Env::get(|mut e| e.symbols.intern_owned(symbol))
    }
}

impl std::fmt::Debug for Symbol {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match Env::get(|e| e.symbols.get(*self)) {
            Some(symbol) => write!(f, "{symbol:?}"),
            None => f.debug_tuple("Symbol").field(&self.0).finish(),
        }
    }
}

impl Display for Symbol {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match Env::get(|e| e.symbols.get(*self)) {
            Some(symbol) => write!(f, "{symbol}"),
            None => write!(f, "<?{}>", self.0),
        }
    }
}

impl Span {
    pub fn intern(span: SpanData) -> Self {
        Env::get(|mut e| e.spans.intern(span))
    }

    #[must_use]
    pub fn join(self, other: Self) -> Self {
        let (self_data, other_data) =
            Env::get(|e| (e.spans.get(self).unwrap(), e.spans.get(other).unwrap()));
        let new_data = self_data.join(&other_data);
        Self::intern(new_data)
    }

    #[must_use]
    pub fn file_id(self) -> usize {
        let data = Env::get(|e| e.spans.get(self).unwrap());
        data.file_id()
    }
}

impl From<Span> for Range<usize> {
    fn from(value: Span) -> Self {
        let data = Env::get(|e| e.spans.get(value).unwrap_or_default());
        data.start()..data.end()
    }
}

impl std::fmt::Debug for Span {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match Env::get(|e| e.spans.get(*self)) {
            Some(span) => write!(f, "{span:?}"),
            None => f.debug_tuple("Span").field(&self.0).finish(),
        }
    }
}

#[derive(Default)]
struct Env {
    symbols: SymbolInterner,
    spans:   SpanInterner,
    ctx:     GlobalCtx,
}

impl Env {
    fn get<T>(f: impl FnOnce(MutexGuard<'_, Self>) -> T) -> T {
        static GLOBAL_DATA: OnceLock<Mutex<Env>> = OnceLock::new();
        f(GLOBAL_DATA.get_or_init(Default::default).lock().unwrap())
    }
}

impl Ty {
    #[must_use]
    pub const fn new_unchecked(kind: &'static TyKind) -> Self {
        Self(kind)
    }

    #[inline]
    pub const fn kind(self) -> &'static TyKind {
        self.0
    }

    pub fn intern(ty: TyKind) -> Self {
        Env::get(|mut e| e.ctx.intern(ty))
    }

    pub fn intern_slice(ty: Vec<Self>) -> TySlice {
        Env::get(|mut e| e.ctx.intern_slice(ty))
    }

    pub fn intern_quant(ty: Vec<TyId>) -> TyQuant {
        Env::get(|mut e| e.ctx.intern_quant(ty))
    }

    #[must_use]
    pub const fn int() -> Self {
        INT
    }

    #[must_use]
    pub const fn bool() -> Self {
        BOOL
    }

    #[must_use]
    pub const fn char() -> Self {
        CHAR
    }

    #[must_use]
    pub const fn real() -> Self {
        REAL
    }

    #[must_use]
    pub const fn empty_slice() -> TySlice {
        EMPTY_SLICE
    }

    #[must_use]
    pub const fn empty_quant() -> TyQuant {
        EMPTY_QUANT
    }
}

struct SymbolInterner {
    indexes: FxHashMap<&'static str, u32>,
    symbols: Vec<&'static str>,
}

macro_rules! default_symbols {
    [$($symbol:literal),+ $(,)?] => {{
        const SYMBOLS_LEN: usize = [$($symbol),+].len();
        const SYMBOLS: [&'static str; SYMBOLS_LEN] = [$($symbol),+];
        let mut indexes = FxHashMap::default();
        indexes.reserve(SYMBOLS_LEN);
        let mut i = 0usize;
        while i < SYMBOLS_LEN {
            #[allow(clippy::cast_possible_truncation)]
            indexes.insert(SYMBOLS[i], i as u32);
            i += 1;
        }
        let symbols = Vec::from(SYMBOLS);
        SymbolInterner {
            indexes,
            symbols,
        }
    }};
}

impl Default for SymbolInterner {
    fn default() -> Self {
        default_symbols![
            "", "+", "-", "/", "*", "^", "^^", "!", "==", "!=", ">", ">=", "<", "<=", "&&", "||",
            ">>", ">>=", "$", ".", "List", "Option", "Result", "Add", "Sub", "Mul", "Div", "Pow",
            "Neg", "Eq", "Cmp", "Number", "And", "Or", "Not", "Nil", "Cons", "Some", "None",
        ]
    }
}

impl SymbolInterner {
    fn get(&self, symbol: Symbol) -> Option<&'static str> {
        self.symbols.get(symbol.0 as usize).copied()
    }

    fn intern(&mut self, symbol: &str) -> Symbol {
        if let Some(idx) = self.indexes.get(symbol) {
            Symbol(*idx)
        } else {
            let symbol = Box::leak(Box::from(symbol));
            let idx = self.symbols.len().try_into().unwrap();
            self.symbols.push(symbol);
            self.indexes.insert(symbol, idx);
            Symbol(idx)
        }
    }

    fn intern_owned(&mut self, symbol: String) -> Symbol {
        if let Some(idx) = self.indexes.get(symbol.as_str()) {
            Symbol(*idx)
        } else {
            let symbol = Box::leak(symbol.into_boxed_str());
            let idx = self.symbols.len().try_into().unwrap();
            self.symbols.push(symbol);
            self.indexes.insert(symbol, idx);
            Symbol(idx)
        }
    }
}

#[derive(Default)]
struct SpanInterner {
    spans: Vec<SpanData>,
}

impl SpanInterner {
    fn get(&self, span: Span) -> Option<SpanData> {
        self.spans.get(span.index()).copied()
    }

    fn intern(&mut self, span: SpanData) -> Span {
        let idx = self.spans.len().try_into().unwrap();
        self.spans.push(span);
        Span::new(idx)
    }
}

#[derive(Default)]
struct GlobalCtx {
    types:       FxHashSet<&'static TyKind>,
    slices:      FxHashSet<&'static [Ty]>,
    quantifiers: FxHashSet<&'static [TyId]>,
}

static INT: Ty = Ty(&TyKind::Int);
static BOOL: Ty = Ty(&TyKind::Bool);
static CHAR: Ty = Ty(&TyKind::Char);
static REAL: Ty = Ty(&TyKind::Real);
static EMPTY_SLICE: TySlice = TySlice(&[]);
static EMPTY_QUANT: TyQuant = TyQuant(&[]);

impl GlobalCtx {
    const fn try_get_primitive(ty: &TyKind) -> Option<Ty> {
        match ty {
            TyKind::Int => Some(INT),
            TyKind::Bool => Some(BOOL),
            TyKind::Char => Some(CHAR),
            TyKind::Real => Some(REAL),
            _ => None,
        }
    }

    fn intern(&mut self, ty: TyKind) -> Ty {
        if let Some(ty) = Self::try_get_primitive(&ty) {
            return ty;
        }
        if let Some(ty) = self.types.get(&ty) {
            return Ty(ty);
        }

        let ty = Box::leak(Box::new(ty));
        self.types.insert(ty);
        Ty(ty)
    }

    fn intern_slice(&mut self, ty: Vec<Ty>) -> TySlice {
        if ty.is_empty() {
            return EMPTY_SLICE;
        }

        if let Some(slice) = self.slices.get(ty.as_slice()) {
            return TySlice(slice);
        }

        let slice = Box::leak(ty.into_boxed_slice());
        self.slices.insert(slice);
        TySlice(slice)
    }

    fn intern_quant(&mut self, ty: Vec<TyId>) -> TyQuant {
        if ty.is_empty() {
            return EMPTY_QUANT;
        }

        if let Some(slice) = self.quantifiers.get(ty.as_slice()) {
            return TyQuant(slice);
        }

        let slice = Box::leak(ty.into_boxed_slice());
        self.quantifiers.insert(slice);
        TyQuant(slice)
    }
}
