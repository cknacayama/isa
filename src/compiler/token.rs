use std::fmt::Display;

use crate::global::Symbol;
use crate::span::Spand;

pub type Token = Spand<TokenKind>;

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum TokenKind {
    Underscore,

    LParen,
    RParen,
    LBrace,
    RBrace,
    LBracket,
    RBracket,

    At,
    Eq,
    Bar,
    DotDot,
    DotDotEq,
    Arrow,
    Rocket,
    Comma,
    Semicolon,
    Colon,
    ColonColon,

    Backslash,

    Integer(i64),
    Real(f64),
    Ident(Symbol),
    Operator(Symbol),
    String(Symbol),
    Char(u8),

    KwTrue,
    KwFalse,

    KwType,
    KwAlias,
    KwLet,
    KwVal,
    KwIn,
    KwModule,
    KwClass,
    KwInstance,
    KwInfix,
    KwInfixl,
    KwInfixr,
    KwPrefix,

    KwSelf,
    KwInt,
    KwBool,
    KwChar,
    KwReal,

    KwMatch,
    KwWith,
    KwIf,
    KwThen,
    KwElse,
}

impl std::hash::Hash for TokenKind {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        core::mem::discriminant(self).hash(state);
    }
}

impl TokenKind {
    #[must_use]
    pub fn keyword(s: &str) -> Self {
        match s {
            "_" => Self::Underscore,
            "let" => Self::KwLet,
            "val" => Self::KwVal,
            "if" => Self::KwIf,
            "then" => Self::KwThen,
            "true" => Self::KwTrue,
            "false" => Self::KwFalse,
            "int" => Self::KwInt,
            "bool" => Self::KwBool,
            "char" => Self::KwChar,
            "real" => Self::KwReal,
            "type" => Self::KwType,
            "alias" => Self::KwAlias,
            "class" => Self::KwClass,
            "instance" => Self::KwInstance,
            "infix" => Self::KwInfix,
            "infixl" => Self::KwInfixl,
            "infixr" => Self::KwInfixr,
            "prefix" => Self::KwPrefix,
            "match" => Self::KwMatch,
            "else" => Self::KwElse,
            "in" => Self::KwIn,
            "with" => Self::KwWith,
            "module" => Self::KwModule,
            "Self" => Self::KwSelf,
            _ => Self::Ident(Symbol::intern(s)),
        }
    }

    pub const fn identifier_character(c: char) -> bool {
        c == '_' || c.is_ascii_alphanumeric()
    }

    pub const fn operator_character(c: char) -> bool {
        matches!(
            c,
            '!' | '?' | '^' | '$' | '%' | '&' | '/' | '*' | '+' | '.' | '<' | '=' | '>' | '|' | '-'
        )
    }

    pub fn operator(op: &str) -> Self {
        match op {
            "=" => Self::Eq,
            "|" => Self::Bar,
            ".." => Self::DotDot,
            "..=" => Self::DotDotEq,
            "->" => Self::Arrow,
            "=>" => Self::Rocket,
            _ => Self::Operator(Symbol::intern(op)),
        }
    }

    pub const fn as_operator(&self) -> Option<Symbol> {
        if let Self::Operator(v) = self {
            Some(*v)
        } else {
            None
        }
    }

    pub const fn as_ident(&self) -> Option<Symbol> {
        if let Self::Ident(v) = self {
            Some(*v)
        } else {
            None
        }
    }
}

impl Display for TokenKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Underscore => write!(f, "`_`"),
            Self::LParen => write!(f, "`(`"),
            Self::RParen => write!(f, "`)`"),
            Self::LBrace => write!(f, "`{{`"),
            Self::RBrace => write!(f, "`}}`"),
            Self::LBracket => write!(f, "`[`"),
            Self::RBracket => write!(f, "`]`"),
            Self::At => write!(f, "`@`"),
            Self::Bar => write!(f, "`|`"),
            Self::Comma => write!(f, "`,`"),
            Self::Semicolon => write!(f, "`;`"),
            Self::Colon => write!(f, "`:`"),
            Self::ColonColon => write!(f, "`::`"),
            Self::DotDot => write!(f, "`..`"),
            Self::DotDotEq => write!(f, "`..=`"),
            Self::Eq => write!(f, "`=`"),
            Self::Arrow => write!(f, "`->`"),
            Self::Rocket => write!(f, "`=>`"),
            Self::Backslash => write!(f, "`\\`"),
            Self::Integer(_) => write!(f, "integer"),
            Self::Real(_) => write!(f, "real"),
            Self::String(_) => write!(f, "string"),
            Self::Ident(_) => write!(f, "identifier"),
            Self::Operator(_) => write!(f, "operator"),
            Self::Char(_) => write!(f, "character"),
            Self::KwTrue => write!(f, "`true`"),
            Self::KwFalse => write!(f, "`false`"),
            Self::KwType => write!(f, "`type`"),
            Self::KwAlias => write!(f, "`alias`"),
            Self::KwLet => write!(f, "`let`"),
            Self::KwVal => write!(f, "`val`"),
            Self::KwIn => write!(f, "`in`"),
            Self::KwClass => write!(f, "`class`"),
            Self::KwInstance => write!(f, "`instance`"),
            Self::KwInfix => write!(f, "`infix`"),
            Self::KwInfixl => write!(f, "`infixl`"),
            Self::KwInfixr => write!(f, "`infixr`"),
            Self::KwPrefix => write!(f, "`prefix`"),
            Self::KwModule => write!(f, "`module`"),
            Self::KwSelf => write!(f, "`Self`"),
            Self::KwInt => write!(f, "`int`"),
            Self::KwBool => write!(f, "`bool`"),
            Self::KwChar => write!(f, "`char`"),
            Self::KwReal => write!(f, "`real`"),
            Self::KwMatch => write!(f, "`match`"),
            Self::KwWith => write!(f, "`with`"),
            Self::KwIf => write!(f, "`if`"),
            Self::KwThen => write!(f, "`then`"),
            Self::KwElse => write!(f, "`else`"),
        }
    }
}
