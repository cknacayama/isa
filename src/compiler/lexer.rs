use std::str::Chars;

use super::error::{LexError, LexErrorKind};
use super::token::{Token, TokenKind};
use crate::global::{Span, Symbol};
use crate::span::SpanData;

const EOF: char = '\0';

#[derive(Debug)]
pub struct Lexer<'a> {
    file_id: usize,
    chars:   Chars<'a>,
    input:   &'a str,
    start:   usize,
    cur:     usize,
}

pub type LexResult<T> = Result<T, LexError>;

impl<'a> Lexer<'a> {
    #[must_use]
    pub fn new(file_id: usize, input: &'a str) -> Self {
        Self {
            file_id,
            input,
            chars: input.chars(),
            start: 0,
            cur: 0,
        }
    }

    fn first(&self) -> char {
        self.chars.clone().next().unwrap_or(EOF)
    }

    fn second(&self) -> char {
        let mut iter = self.chars.clone();
        iter.next();
        iter.next().unwrap_or(EOF)
    }

    fn is_eof(&self) -> bool {
        self.chars.as_str().is_empty()
    }

    fn bump(&mut self) -> Option<char> {
        self.chars.next().inspect(|c| self.cur += c.len_utf8())
    }

    fn bump_twice(&mut self) -> Option<(char, char)> {
        let c1 = self.chars.next()?;
        self.cur += c1.len_utf8();
        let c2 = self.chars.next()?;
        self.cur += c2.len_utf8();
        Some((c1, c2))
    }

    fn match_cur(&mut self, to_match: char) -> bool {
        if self.first() == to_match {
            self.bump();
            true
        } else {
            false
        }
    }

    fn eat_while<P>(&mut self, pred: P)
    where
        P: Fn(char) -> bool,
    {
        while pred(self.first()) && !self.is_eof() {
            self.bump();
        }
    }

    fn make_token(&self, kind: TokenKind) -> Token {
        let span = SpanData::new(self.file_id, self.start, self.cur);
        let span = Span::intern(span);
        Token::new(kind, span)
    }

    fn make_err(&self, kind: LexErrorKind) -> LexError {
        let span = SpanData::new(self.file_id, self.start, self.cur);
        let span = Span::intern(span);
        LexError::new(kind, span)
    }

    fn skip_whitespace(&mut self) {
        while !self.is_eof() {
            match self.first() {
                '/' if self.second() == '/' => {
                    self.bump_twice();
                    self.eat_while(|c| c != '\n');
                }
                c if c.is_ascii_whitespace() => {
                    self.bump();
                }
                _ => break,
            }
        }
    }

    fn number(&mut self) -> Token {
        self.eat_while(|c| c.is_ascii_digit());

        if self.first() == '.' && self.second().is_ascii_digit() {
            self.bump_twice();
            self.eat_while(|c| c.is_ascii_digit());
            let s = &self.input[self.start..self.cur];
            self.make_token(TokenKind::Real(s.parse().unwrap()))
        } else {
            let s = &self.input[self.start..self.cur];
            self.make_token(TokenKind::Integer(s.parse().unwrap()))
        }
    }

    fn character(&mut self) -> LexResult<Token> {
        let c = self
            .bump()
            .ok_or_else(|| self.make_err(LexErrorKind::UnterminatedChar))?;

        let c = match c {
            '\\' => {
                let c = self
                    .bump()
                    .ok_or_else(|| self.make_err(LexErrorKind::UnterminatedString))?;
                Self::escape(c).ok_or_else(|| {
                    self.eat_while(|c| c != '"');
                    self.make_err(LexErrorKind::UnrecognizedEscape(c))
                })? as u8
            }
            '\'' => return Err(self.make_err(LexErrorKind::EmptyChar)),
            _ if c.is_ascii() => c as u8,
            _ => return Err(self.make_err(LexErrorKind::InvalidChar(c))),
        };

        if self.match_cur('\'') {
            Ok(self.make_token(TokenKind::Char(c)))
        } else {
            Err(self.make_err(LexErrorKind::UnterminatedChar))
        }
    }

    const fn escape(c: char) -> Option<char> {
        match c {
            'n' => Some('\n'),
            'r' => Some('\r'),
            't' => Some('\t'),
            '0' => Some('\0'),
            '\\' => Some('\\'),
            '\'' => Some('\''),
            '\"' => Some('\"'),
            _ => None,
        }
    }

    fn string(&mut self) -> LexResult<Token> {
        let mut s = String::new();

        while !self.is_eof() && self.first() != '"' && self.first() != '\n' {
            let c = self.bump().unwrap();
            if c == '\\' {
                let c = self
                    .bump()
                    .ok_or_else(|| self.make_err(LexErrorKind::UnterminatedString))?;
                let c = Self::escape(c).ok_or_else(|| {
                    self.eat_while(|c| c != '"');
                    self.make_err(LexErrorKind::UnrecognizedEscape(c))
                })?;
                s.push(c);
            } else {
                s.push(c);
            }
        }

        if !self.match_cur('"') {
            return Err(self.make_err(LexErrorKind::UnterminatedString));
        }

        let kind = TokenKind::String(Symbol::intern_owned(s));

        Ok(self.make_token(kind))
    }

    fn identifier_or_keyword(&mut self) -> Token {
        self.eat_while(TokenKind::identifier_character);
        let s = &self.input[self.start..self.cur];
        let kind = TokenKind::keyword(s);
        self.make_token(kind)
    }

    fn operator(&mut self) -> Token {
        self.eat_while(TokenKind::operator_character);
        let s = &self.input[self.start..self.cur];
        let kind = TokenKind::operator(s);
        self.make_token(kind)
    }

    pub fn next_token(&mut self) -> Option<LexResult<Token>> {
        self.skip_whitespace();

        self.start = self.cur;
        let c = self.bump()?;

        macro_rules! token {
            ($tk:ident) => {{ Some(Ok(self.make_token(TokenKind::$tk))) }};
            ($tk:ident, $c2:expr => $tk2:ident) => {{
                if self.match_cur($c2) {
                    Some(Ok(self.make_token(TokenKind::$tk2)))
                } else {
                    Some(Ok(self.make_token(TokenKind::$tk)))
                }
            }};
        }

        match c {
            '@' => token!(At),
            '(' => token!(LParen),
            ')' => token!(RParen),
            '[' => token!(LBracket),
            ']' => token!(RBracket),
            '{' => token!(LBrace),
            '}' => token!(RBrace),
            ',' => token!(Comma),
            '\\' => token!(Backslash),
            ';' => token!(Semicolon),
            ':' => token!(Colon, ':' => ColonColon),
            '0'..='9' => Some(Ok(self.number())),
            '\'' => Some(self.character()),
            '"' => Some(self.string()),
            '_' | 'a'..='z' | 'A'..='Z' => Some(Ok(self.identifier_or_keyword())),
            c if TokenKind::operator_character(c) => Some(Ok(self.operator())),
            _ => Some(Err(self.make_err(LexErrorKind::InvalidChar(c)))),
        }
    }
}

impl Iterator for Lexer<'_> {
    type Item = LexResult<Token>;

    fn next(&mut self) -> Option<Self::Item> {
        self.next_token()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    macro_rules! assert_tk {
        ($input:literal) => {
            insta::assert_debug_snapshot!(Lexer::new(0, $input).next())
        };
    }

    macro_rules! assert_all {
        ($input:literal) => {
            insta::assert_debug_snapshot!(Lexer::new(0, $input).collect::<Vec<_>>())
        };
    }

    #[test]
    fn numbers() {
        assert_tk!("0000");
        assert_tk!("0120");
        assert_tk!("12");
        assert_tk!("4.0");
        assert_tk!("5.0");
        assert_tk!("1.2");
        assert_tk!("111.2");
    }

    #[test]
    fn keyword() {
        assert_tk!("_");
        assert_tk!("let");
        assert_tk!("val");
        assert_tk!("if");
        assert_tk!("then");
        assert_tk!("true");
        assert_tk!("false");
        assert_tk!("int");
        assert_tk!("bool");
        assert_tk!("char");
        assert_tk!("real");
        assert_tk!("type");
        assert_tk!("alias");
        assert_tk!("class");
        assert_tk!("instance");
        assert_tk!("infix");
        assert_tk!("infixl");
        assert_tk!("infixr");
        assert_tk!("prefix");
        assert_tk!("match");
        assert_tk!("else");
        assert_tk!("in");
        assert_tk!("with");
        assert_tk!("module");
        assert_tk!("Self");
    }

    #[test]
    fn ident() {
        assert_tk!("_i");
        assert_tk!("letf");
        assert_tk!("vals");
        assert_tk!("ifs");
        assert_tk!("thenz");
        assert_tk!("true10");
        assert_tk!("false18");
        assert_tk!("aint");
        assert_tk!("lbool");
        assert_tk!("_char");
        assert_tk!("real_");
        assert_tk!("Type");
        assert_tk!("aliass");
        assert_tk!("classes");
        assert_tk!("_8instance");
        assert_tk!("_01infix");
        assert_tk!("_01infixl");
        assert_tk!("_01infixr");
        assert_tk!("mengoprefix");
        assert_tk!("amatch");
        assert_tk!("belse");
        assert_tk!("cin");
        assert_tk!("dwith");
        assert_tk!("emodule");
        assert_tk!("self");
    }

    #[test]
    fn operator() {
        assert_tk!("+");
        assert_tk!("-");
        assert_tk!("/");
        assert_tk!("*");
        assert_tk!("^");
        assert_tk!("^^");
        assert_tk!("!");
        assert_tk!("==");
        assert_tk!("!=");
        assert_tk!(">");
        assert_tk!(">=");
        assert_tk!("<");
        assert_tk!("<=");
        assert_tk!("&&");
        assert_tk!("||");
        assert_tk!(">>");
        assert_tk!(">>=");
        assert_tk!("$");
        assert_tk!(".");
    }

    #[test]
    fn comment() {
        let input = "// mengo mengo mengo mengo 123 && Ã  aa^`";
        let mut lexer = Lexer::new(0, input);
        assert!(lexer.next().is_none());
    }

    #[test]
    fn comment_and_whitespace() {
        let input = "\n\n\n\n// mengo mengo mengo mengo 123 && a aa^`\t\t\t\t\n//foo bar baz";
        let mut lexer = Lexer::new(0, input);
        assert!(lexer.next().is_none());
    }

    #[test]
    fn spaced_tokens() {
        let input =
            "\n\n12 +\n\n// mengo mengo mengo mengo 123 && a aa^`\t\t\t\t\n//foo bar baz\n\n   13";
        let mut lexer = Lexer::new(0, input);
        insta::assert_debug_snapshot!(lexer.next());
        insta::assert_debug_snapshot!(lexer.next());
        insta::assert_debug_snapshot!(lexer.next());
        insta::assert_debug_snapshot!(lexer.next());
    }

    #[test]
    fn chars() {
        assert_tk!(r"'a'");
        assert_tk!(r"'\n'");
        assert_tk!(r"'\t'");
        assert_tk!(r"'\r'");
        assert_tk!(r"'\\'");
        assert_tk!(r"'\''");
        assert_tk!(r#"'\"'"#);
        assert_tk!(r"'\0'");
        assert_tk!(r"'1'");
    }

    #[test]
    fn strings() {
        assert_tk!(r#""foo""#);
        assert_tk!(r#""foo\n""#);
        assert_tk!(r#""bar\r""#);
        assert_tk!(r#""ba5\t""#);
        assert_tk!(r#""ba3\"""#);
        assert_tk!(r#""ba2\0a""#);
    }

    #[test]
    fn delimiters() {
        assert_tk!("(");
        assert_tk!(")");
        assert_tk!("{");
        assert_tk!("}");
        assert_tk!("[");
        assert_tk!("]");
        assert_tk!(":");
        assert_tk!("::");
        assert_tk!(";");
        assert_tk!("@");
        assert_tk!("->");
        assert_tk!("=>");
        assert_tk!("\\");
        assert_tk!(",");
    }

    #[test]
    fn invalid_errors() {
        assert_tk!(r"Î»");
        assert_all!(r"Î» Î» Î»");
        assert_all!(r"Î» ba Î»");
    }

    #[test]
    fn char_errors() {
        assert_tk!("''");
        assert_tk!(r"'\z'");
        assert_tk!(r"'Î»'");
        assert_tk!(r"'aa'");
    }

    #[test]
    fn string_errors() {
        assert_tk!(r#""foo"#);
        assert_tk!(r#""foo\n"#);
        assert_tk!(r#""bar\"#);
        assert_tk!(
            r#""ba3\"
            ""#
        );
        assert_tk!(r#""ba5\c""#);
    }

    #[test]
    fn lists() {
        assert_all!("[]");
        assert_all!("[1,2,3]");
        assert_all!("[a,b,c,d]");
        assert_all!("[['a'],['b'],['c'],['d']]");
        assert_all!("[['a'],['b'],['c'],['d'],]");
    }

    #[test]
    fn tuples() {
        assert_all!("()");
        assert_all!("(1,2,3)");
        assert_all!("(a,b,c,d)");
        assert_all!("(('a'),('b'),('c'),('d'))");
        assert_all!("(('a'),('b'),('c'),('d'),)");
    }

    #[test]
    fn lets() {
        assert_all!("let a = 10 in 10");
        assert_all!("let a c = 10 in let b = a in b");
    }

    #[test]
    fn paths() {
        assert_all!("a::b");
        assert_all!("a::b::c");
        assert_all!("a::b::c a");
        assert_all!("a::b::c a::T");
        assert_all!("a::b::c C::a::T b");
    }

    #[test]
    fn prefix() {
        assert_all!("-1");
        assert_all!("(!false)");
        assert_all!("-a");
        assert_all!(r#"-"string""#);
        assert_all!(r#"-'c'"#);
        assert_all!(r#"- - - - - - - + - + 1"#);
    }

    #[test]
    fn infix() {
        assert_all!("1 + 1");
        assert_all!("a + 1 + 1");
        assert_all!(r#""string" + (1 * 1)"#);
        assert_all!("a + ((c ^^ []) >>= 'd')");
        assert_all!("1 * 2 * 3 + ((() & (1) ^^ []) >>= 'd')");
    }

    #[test]
    fn lambdas() {
        assert_all!(r"\x -> x");
        assert_all!(r"\(a,b) -> a + b");
        assert_all!(r"\_ -> 10 + (\x -> x) 10");
    }

    #[test]
    fn calls() {
        assert_all!("Some x");
        assert_all!(r"map (\x -> x + 1) [1,2,3]");
        assert_all!(r"(>>=) (None) \x -> return x");
    }

    #[test]
    fn matches() {
        assert_all!("match 10 with 0.. -> 10, ..0 -> -10");
        assert_all!("match (1,2) with (0.., ..0) -> 10, (..0, 0..) -> -10");
        assert_all!("match [10] with [] -> match [] with _ -> 10,, [a]b -> a+b");
    }

    #[test]
    fn ifs() {
        assert_all!("if true then true else false");
        assert_all!("if false then if true then false else true else false");
        assert_all!("if !false then !true else if false then true else false");
    }

    #[test]
    fn let_stmts() {
        assert_all!("let a = 10;");
        assert_all!("let fib n = match n with ..2 -> n, _ -> fib (n - 1) + fib (n - 2)");
        assert_all!("let foo (a,b,c) = a + b - c;");
        assert_all!("let bar _ _ _ = true;");
        assert_all!("let mengo _ _c _ = true;");
    }

    #[test]
    fn program() {
        assert_all!(
            r#"
module main

let a = 10;
let c = 1234;
let b = a + b * c;
// comment abc &&
(&&) && (&&);
match c with _ -> 10;
match c with ..10 -> 6.2, 10.. -> 3.14;
"#
        )
    }
}
