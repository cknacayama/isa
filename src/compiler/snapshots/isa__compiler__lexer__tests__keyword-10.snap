---
source: src/compiler/lexer.rs
expression: "Lexer :: new(0, \"char\").next()"
---
Some(
    Ok(
        Spand {
            data: KwChar,
            span: SpanData { file_id: 0, lo: 0, hi: 4 },
        },
    ),
)
